compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -9343,4 +9375,342 @@ private STNode parseIntersectionTypeDescriptor(STNode leftTypeDesc, ParserRuleCo
+
+        endContext();
+        return STNodeFactory.createTypedBindingPatternNode(typeDesc, bindingPattern);
+    }
+
+    private STNode parseArrayTypeDescOrListBindingPattern(STNode typeDesc) {
+
+        STNode arrayDescOrListBindingPattern = parseListBindingPattern();
+
+        if (isListBindingPatternDefinitively(arrayDescOrListBindingPattern)) { // ambiguous means T[a]
+            return STNodeFactory.createTypedBindingPatternNode(typeDesc,
+                    cleanListBindingPattern(arrayDescOrListBindingPattern));
+        }
+
+        if (isFollowTypedBindingPattern(peek().kind)) {
+            return STNodeFactory.createTypedBindingPatternNode(typeDesc,
+                    cleanListBindingPattern(arrayDescOrListBindingPattern));
+        }
+
+        typeDesc = mergeTypeDescAndListBindingPattern(typeDesc, arrayDescOrListBindingPattern);
+        if (peek().kind == SyntaxKind.OPEN_BRACKET_TOKEN) {
+            //this means arrayDescOrBLP is a array type desc
+            return parseArrayTypeDescOrListBindingPattern(typeDesc);
+        } else {
+            return STNodeFactory.createTypedBindingPatternNode(typeDesc,
+                    parseBindingPattern());
+        }
+
+
+    }
+
+    private STNode mergeTypeDescAndListBindingPattern(STNode typeDesc,
+                                                      STNode bindingPattern) {
+        STListBindingPatternNode listBindingPattern = (STListBindingPatternNode) bindingPattern;
+        STNodeList childArray = (STNodeList) listBindingPattern.bindingPatterns;
+        STNode child = childArray.childInBucket(0);
+        if (child.kind == SyntaxKind.CAPTURE_BINDING_PATTERN) {
+            child = ((STCaptureBindingPatternNode) child).variableName;
+        }
+        return STNodeFactory.createIndexedExpressionNode(typeDesc, listBindingPattern.openBracket,
+                child,
+                listBindingPattern.closeBracket);
+    }
+
+    private STNode cleanListBindingPattern(STNode bindingPattern) {
+        STListBindingPatternNode listBindingPattern = (STListBindingPatternNode) bindingPattern;
+        STNodeList childArray = (STNodeList) listBindingPattern.bindingPatterns;
+        int numberOfChildren = childArray.bucketCount();
+
+        ArrayList<STNode> cleanedChildren = new ArrayList<>();
+        STNode child = null;
+        for (int i = 0; i < numberOfChildren; i++) {
+            child = childArray.childInBucket(i);
+            switch (child.kind) {
+                case DECIMAL_INTEGER_LITERAL:
+                case HEX_INTEGER_LITERAL:
+                case ASTERISK_TOKEN:
+                    this.errorHandler.reportInvalidNode(null,
+                            "invalid Node " + child.kind + " in list-binding-pattern");
+                    cleanedChildren.add(STNodeFactory.createMissingToken(SyntaxKind.IDENTIFIER_TOKEN));
+                    break;
+                default:
+                    cleanedChildren.add(child);
```

Comment:
Regardless whether its a valid child or an invalid child, we need to keep it. That is required to regenerate the source-code using the syntax tree.

So we can simply rename this method as "validateListBindingPattern", and only log an error for invalid children, but let them stay.